File: client/src/game/trees/willow.ts
// Weeping willow. The cascade IS the tree: a filled curtain whose top is the
// crown dome and whose hem is scalloped by the strand ends, strands raked down
// over it from right across the crown, and a cave left in the middle of the hem
// so the trunk shows through the way it does under a real willow.
//
//   form 0 — great dome: wider than tall, hem almost on the grass.
//   form 1 — young: narrower, crown held higher, far more trunk showing.

import { PIXEL, SCALE } from '../constants'
import { disc, jitter, trunkScale } from './variance'
import type { Ctx, DrawArgs } from './types'

export const WILLOW_TONES = {
  trunk: '#7A5A12',
  dark: '#548F4E',
  mid: '#6FAB5C',
  light: '#97C877',
}

/** Hem segments; more would blur the scallops at this scale, fewer look cut. */
const SEGS = 9

export function drawWillow(
  ctx: Ctx,
  px: number,
  py: number,
  { rng, form, ink }: DrawArgs,
): void {
  const t = WILLOW_TONES
  const j = jitter(rng)
  const foot = py + PIXEL * 2
  const cx = px + PIXEL + j.lean * 0.4
  const great = form === 0
  const half = PIXEL * (great ? 1.0 : 0.74) * j.w
  const hemY = foot - PIXEL * (great ? 0.02 : 0.5) * j.h
  // The curtain's depth is the roll; the trunk in the cave stays much the same.
  const crownY = hemY - PIXEL * (great ? 1.52 : 1.12) * j.h
  const strands = great ? 22 + Math.round(j.d * 8) : 15 + Math.round(j.d * 6)
  const ts = trunkScale(j)
  const trunkH = PIXEL * (great ? 0.9 : 1.2) * ts

  // Hem: deepest in the middle, drawing in at the rim, with a cave at the centre.
  const hem: { x: number; y: number }[] = []
  for (let i = 0; i <= SEGS; i++) {
    const f = i / SEGS
    const bell = Math.sin(f * Math.PI)
    const cave = Math.exp(-Math.pow((f - 0.5) / 0.13, 2))
    hem.push({
      x: cx + (f - 0.5) * 2 * half,
      y:
        hemY -
        PIXEL * (0.62 - bell * 0.54 + cave * 0.5) * j.h -
        rng() * PIXEL * 0.12,
    })
  }

  // Trunk first: the cave in the hem leaves its lower half showing, and the
  // curtain covers where it ends — a trunk painted on top reads as a cut stump.
  ctx.fillStyle = ink(t.trunk)
  ctx.beginPath()
  ctx.moveTo(cx - PIXEL * 0.24 * ts, foot)
  ctx.lineTo(cx - PIXEL * 0.12 * ts, foot - trunkH)
  ctx.lineTo(cx + PIXEL * 0.12 * ts, foot - trunkH)
  ctx.lineTo(cx + PIXEL * 0.24 * ts, foot)
  ctx.closePath()
  ctx.fill()

  ctx.fillStyle = ink(t.dark)
  ctx.beginPath()
  ctx.moveTo(hem[0].x, hem[0].y)
  ctx.lineTo(cx - half, crownY + PIXEL * 0.2)
  ctx.quadraticCurveTo(
    cx - half * 0.98,
    crownY - PIXEL * 0.6,
    cx - half * 0.46,
    crownY - PIXEL * 0.76,
  )
  ctx.quadraticCurveTo(
    cx - half * 0.04,
    crownY - PIXEL * 0.92,
    cx + half * 0.44,
    crownY - PIXEL * 0.72,
  )
  ctx.quadraticCurveTo(
    cx + half * 0.98,
    crownY - PIXEL * 0.54,
    cx + half,
    crownY + PIXEL * 0.2,
  )
  ctx.lineTo(hem[SEGS].x, hem[SEGS].y)
  for (let i = SEGS; i > 0; i--) {
    const a = hem[i]
    const b = hem[i - 1]
    ctx.quadraticCurveTo(
      (a.x + b.x) / 2,
      Math.max(a.y, b.y) + PIXEL * 0.22,
      b.x,
      b.y,
    )
  }
  ctx.closePath()
  ctx.fill()

  // Strands falling from right across the crown — not converging on the middle,
  // which is what makes a weeping tree read as a mop.
  ctx.lineCap = 'round'
  for (let i = 0; i < strands; i++) {
    const f = (i + 0.5) / strands
    const bell = Math.sin(f * Math.PI)
    const cave = Math.exp(-Math.pow((f - 0.5) / 0.11, 2))
    const sx = cx + (f - 0.5) * 2 * half * 0.8
    const sy = crownY - PIXEL * (0.18 + bell * 0.46) * j.h
    const ex = cx + (f - 0.5) * 2 * half * (0.9 + rng() * 0.12)
    const ey =
      hemY -
      PIXEL * (0.54 - bell * 0.5 + cave * 0.62) * j.h -
      rng() * PIXEL * 0.22
    ctx.strokeStyle = ink(i % 4 === 0 ? t.light : t.mid)
    ctx.lineWidth = SCALE * (i % 4 === 0 ? 0.9 : 0.7)
    ctx.beginPath()
    ctx.moveTo(sx, sy)
    ctx.quadraticCurveTo(sx + (ex - sx) * 0.9, sy + (ey - sy) * 0.5, ex, ey)
    ctx.stroke()
  }

  // Light on the crown's shoulder.
  ctx.fillStyle = ink(t.light)
  disc(ctx, cx - half * 0.42, crownY - PIXEL * 0.46, PIXEL * 0.3 * j.w)
  disc(ctx, cx + half * 0.06, crownY - PIXEL * 0.6, PIXEL * 0.26 * j.w)
  disc(ctx, cx + half * 0.56, crownY - PIXEL * 0.36, PIXEL * 0.2 * j.w)
}

File: 3d/blender/trees/willow.py
"""Weeping willow — the cascade IS the tree.

Ports `client/src/game/trees/willow.ts`:

  form 0 — great dome: wider than tall, hem almost on the grass.
  form 1 — young:      narrower, crown held higher, far more trunk showing.

In 2D the cascade is a filled curtain with a scalloped hem and a cave left in
the middle so the trunk shows through. In 3D the equivalent is a crown dome with
strands raked down around it — tapered tubes, the same technique as the koala's
tail, arranged in a ring with a gap left at the front so the trunk still reads.

Run from the package root:
    blender --background --python blender/trees/willow.py
"""

import math
import os
import sys

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
import lib  # noqa: E402

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
import _common as tc  # noqa: E402

TONES = {"trunk": "#7A5A12", "dark": "#548F4E", "mid": "#6FAB5C", "light": "#97C877"}


def build(form):
    lib.reset_scene()
    col = lib.collection(f"tree_willow_{form}")
    rand = tc.rng(0x3111 + form)

    bark = lib.clay("willow.trunk", TONES["trunk"], rough=0.74)
    dark = lib.clay("willow.dark", TONES["dark"], rough=0.68)
    mid = lib.clay("willow.mid", TONES["mid"], rough=0.66)
    light = lib.clay("willow.light", TONES["light"], rough=0.64)

    dome = form == 0
    trunk_h = 0.86 if dome else 1.24
    crown_z = trunk_h + (0.40 if dome else 0.34)
    radius = 1.16 if dome else 0.82
    height = 0.52 if dome else 0.48
    # how far the strands fall — the great dome's hem nearly reaches the grass
    fall = crown_z - (0.14 if dome else 0.56)
    strands = 26 if dome else 20

    lib.add(col, tc.trunk(trunk_h, 0.100, 0.064), bark).name = "trunk"

    def tone(obj, mat):
        obj.data.materials.clear()
        obj.data.materials.append(mat)
        for p in obj.data.polygons:
            p.use_smooth = True
        return obj

    parts = []

    # the crown dome the strands hang from
    for p in tc.blob_crown(rand, (0, 0, crown_z), radius * 0.86, height, count=6):
        parts.append(tone(p, dark))

    # strands raked down around the crown. A gap is left at the front (+X) so
    # the trunk shows through, the way the 2D leaves a cave in the hem.
    for i in range(strands):
        t = i / strands
        ang = t * math.tau
        # skip the front arc
        if abs(((ang + math.pi) % math.tau) - math.pi) < 0.55:
            continue
        reach = radius * (0.72 + rand() * 0.32)
        drop = fall * (0.70 + rand() * 0.38)
        top = crown_z + height * 0.10
        bottom = max(0.06, top - drop)
        x, y = math.cos(ang) * reach, math.sin(ang) * reach
        strand = tc.tube(
            f"strand_{i}",
            [(0.0, top, 0.052),
             (reach * 0.18, top - drop * 0.42, 0.040),
             (reach * 0.24, bottom, 0.020)],
            bevel_res=1, res_u=4,
        )
        strand.rotation_euler = (0, 0, ang)
        strand.location = (x * 0.72, y * 0.72, 0.0)
        parts.append(tone(strand, light if i % 3 == 0 else mid))

    crown = tc.merge(parts, "crown")
    for c in crown.users_collection:
        c.objects.unlink(crown)
    col.objects.link(crown)
    crown.name = "crown"

    lib.seat_on_ground(col)
    return col


if __name__ == "__main__":
    for form in (0, 1):
        lib.export(build(form), f"trees/willow_{form}.glb", budget=6000)
